Android client

lfb-anssi opened the pull request:
This pull request ends the Android client part of the protocol.

It also adds support for PCR extension on both server and client side:
If the --pcr-extend option is used during the enrollment (with --enroll),
a secret will be generated on client side, and sent back on each attesatation.
The server will then extend the 9th PCR with it.
This allows the user to base disk decryption of his machine on the attestation response.

Commit:
Fix fragment names

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Create a new datastructure to send byte arrays

As Bytestring type is not well handled for several
CBOR implementations, creating a struct to wrap them
makes things easier.

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Remove usage indications at runtime

Also fix nonce encoding on reception

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
update serialization plugins

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Continue the protocol implementation until quote signature validation

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Add .gitignore file for the client go code

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Fix issue #10, remove the registration part

- Fix the issue that makes the server crash if a
virtual TPM was used. Now we check that the EK
certificate exists before dereferencing it.

- The registration was made on successful EK exchange.
If an error arised after that, invalid devices would have
been stored on both sides.
This part has gone, and now we register both devices on both
side at the end of the protocol: in the 'response' protocol
function.

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Add support for PCR extension on ultrablue success

If the --pcr-extend option is specified at enroll time,
the client will generate a secret and send it back for
each successful attestation.
If that secret is received by the server, it will
extend the 9th PCR with.

Updates testbed according to those changes, and add
TPM2 sealing documentation.

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Disconnect BLE on protocol end

Disconnect BLE on protocol end to avoid pending connections,
e.g. on failure.
Also introduce the start protocol method, to make the code
more clear.

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
extract read/write messages to strings file

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

nbouchinet-anssi commented on server/main.go:
```diff
@@ -16,9 +16,10 @@ import (
-	loglevel = flag.Int("loglevel", 0, "Indicates the level of logging, 0 is the minimum, 3 is the maximum")
-	mtu      = flag.Int("mtu", 500, "Set a custom MTU, which is basically the max size of the BLE packets")
+	enroll       = flag.Bool("enroll", false, "Must be true for a first time attestation")
+	loglevel     = flag.Int("loglevel", 1, "Indicates the level of logging, 0 is the minimum, 3 is the maximum")
```
Why did you changed the default loglevel ?

buckwell-anssi replied:
I changed the loglevel to 1 by default so that the user is aware of the attestation state. If using ultrablue for disk decryption, it provides visual information so that the user knows what is going on.

nbouchinet-anssi commented on server/main.go:
```diff
@@ -16,9 +16,10 @@ import (
+	enroll       = flag.Bool("enroll", false, "Must be true for a first time attestation")
+	loglevel     = flag.Int("loglevel", 1, "Indicates the level of logging, 0 is the minimum, 3 is the maximum")
+	mtu          = flag.Int("mtu", 500, "Set a custom MTU, which is basically the max size of the BLE packets")
+	pcrextend    = flag.Bool("pcr-extend", false, "Extend the 9th PCR with the verifier secret, on attestation success")
```
Typo, remove the ending comma

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -82,26 +84,37 @@ func recvMsg[T any](obj *T, ch chan []byte) error {
 */
 
-// EK contains a RSA public key and an optional certificate.
+// RegistrationData contains the TPM's endorsement RSA public keywith
```
Typo "keywith"

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -82,26 +84,37 @@ func recvMsg[T any](obj *T, ch chan []byte) error {
 }
 
-func registration(ch chan []byte, tpm *attest.TPM) error {
+func ekenroll(ch chan []byte, tpm *attest.TPM) error {
```
Can you be consistent in your function naming convention ?

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -121,41 +134,32 @@ func registration(ch chan []byte, tpm *attest.TPM) error {
 	return nil
 }
 
+
```
Typo, remove empty line

nbouchinet-anssi commented on server/protocol.go:
```diff
@@ -82,26 +84,37 @@ func recvMsg[T any](obj *T, ch chan []byte) error {
+	PCRExtend bool   // Wether or not PCR_EXTENSION_INDEX must be extended on attestation success
+}
+
+type Bytestring struct {
```
Please comment on your Bytestring encoding workaround

kerneis-anssi replied:
This conservation is marked as resolved but I can't find a comment.

nbouchinet-anssi commented on server/testbed/README.md:
````diff
@@ -24,4 +25,17 @@ following dracut command:
+systemd-cryptenroll --tpm2-device=auto --tpm2-pcrs=9 /dev/sdx
+```
+
+Re-generate your initramfs in order to include the ultrablue dracut module:
````
and the crypt module

nbouchinet-anssi commented on server/utils.go:
```diff
@@ -28,6 +29,20 @@ func getTPMRandom(size uint16) ([]byte, error) {
+	}
+	defer rwc.Close()
+
+	err = tpm2.PCREvent(rwc, tpmutil.Handle(index), secret)
```
Please add a TODO to specify that the PCREvent do not feed the eventlog

buckwell-anssi replied:
Done, I also added a note, if you want to check.

nbouchinet-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,154 @@
+	keys, unless we implement others methods, for each specific
+	algorithm.
+*/
+func createRSAPublicKey(publicBytes []byte, exponent int) rsa.PublicKey {
```
createRSAPublicKey function name make me think that you generate a publickey, which is not the case, can you rename it to something like rebuildRSAPublicKey ?

nbouchinet-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+/*
+	createRSAPublicKey deconstructs a crypto.PublicKey in a
```
You reconstruct here

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -1,25 +1,55 @@
-const val REGISTRATION_READ = 0
-const val REGISTRATION = 1
+const val EK_READ = 0
+const val EK_DECODE = 1
```
Can you change those to ENROLL_{READ,DECODE} to be consistent with AUTHENTICATION_{READ,DECODE} ?

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -30,36 +60,166 @@ const val CREDENTIAL_ACTIVATION_ASSERT = 6
-                // When encryption will be implemented, we'll first need to decrypt the nonce here.
-                writeMsg("Nonce", message)
+                if (message.size != 24) {
+                    logger?.push(CLog("Invalid nonce length. Make sure you ran the attestation server without the --enroll flag.", false))
```
Can you log a more user friendly comment here, the user should understand that he is running the android client in the attestation whereas the server is running in enrollment mode instead of attestation mode.

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -30,36 +60,166 @@ const val CREDENTIAL_ACTIVATION_ASSERT = 6
             AUTHENTICATION -> {
-                // When encryption will be implemented, we'll first need to decrypt the nonce here.
-                writeMsg("Nonce", message)
+                if (message.size != 24) {
+                    logger?.push(CLog("Invalid nonce length. Make sure you ran the attestation server without the --enroll flag.", false))
```
Seems hackish to me, is there no other way to ensure the user is not misusing the protocol ?

buckwell-anssi replied:
I can do that by wrapping the CBOR decoding into a try/catch block, so that it will throw if a wrong data structure is sent, but I'll still have to make the detection here as there's only one Bluetooth characteristic, thus the server 'chooses' what to send, and the only thing we can do is check that it matches what we expect to receive.
It happens that if the wrong mode is choosen, we will figure it at this point.

Do you prefer the try/catch way?

nbouchinet-anssi replied:
I'd prefer the later as it will catch any kind of protocol misuse and not only the invalid nonce one.

kerneis-anssi replied:
Agreed, catch all errors, and still try to provide a helpful error message ("error blabla. Check that the server and client are both running in the same mode.").

Another option is to redesign the message format to prefix them with a message type, and check that we get message of the expected type at each step. The technique is known as TLV and is a good design approach to evolve your protocol later: https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value

Let's not redesign at this stage, but feel free to create an issue to track this enhancement.

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -30,36 +60,166 @@ const val CREDENTIAL_ACTIVATION_ASSERT = 6
+    private var encodedPCRs: ByteArray? = null
+    private var attestationResponse: AttestationResponse? = null
+
+
+    fun start() {
         resume()
     }
 
     @OptIn(ExperimentalSerializationApi::class)
     private fun resume() {
         when (state) {
-            REGISTRATION_READ -> readMsg("EkPub and EkCert")
-            REGISTRATION -> {
-                val key = Cbor.decodeFromByteArray<EKModel>(message)
-                device.name = "device" + device.uid
-                device.ekn = key.N
-                device.eke = key.E.toInt()
-                device.ekcert = key.Cert
-                activity.viewModel.insert(device)
-                writeMsg("registration confirmation", byteArrayOf(-10)) // Registration success, CBOR encoded
+            EK_READ -> readMsg(activity.getString(R.string.ek_pub_cert))
+            EK_DECODE -> {
+                ek = Cbor.decodeFromByteArray(message)
+                state += 1
+                resume()
             }
-            AUTHENTICATION_READ -> readMsg("authentication nonce")
+            AUTHENTICATION_READ -> readMsg(activity.getString(R.string.auth_nonce))
             AUTHENTICATION -> {
-                // When encryption will be implemented, we'll first need to decrypt the nonce here.
-                writeMsg("Nonce", message)
+                if (message.size != 24) {
+                    logger?.push(CLog("Invalid nonce length. Make sure you ran the attestation server without the --enroll flag.", false))
+                    return
+                }
+                val authNonce = Cbor.decodeFromByteArray<ByteArrayModel>(message)
+                //TODO: When encryption will be implemented, we'll need to decrypt the nonce here.
+                val encodedAuthNonce = Cbor.encodeToByteArray(authNonce)
+                writeMsg(activity.getString(R.string.decrypted_auth_nonce), encodedAuthNonce)
+            }
```
Why would we implement the authentication part if it is not used with encryption communication ?

buckwell-anssi replied:
I know you're not found of it, but as TODO are left in the code, and the encryption part should be implemented soon, I think it's worth to keep them, as it'll make it clear where to implement the authentication for the one that will handle it.

nbouchinet-anssi replied:
I'd prefer you move thoses  to another "encrypted-communication" branch for a later PR as we may tag a first version after this one is cleaned-up and merged. What's your opinion @kerneis-anssi ?
I'm not fond of the idea to have partially-used protocols chunks.

nbouchinet-anssi replied:
I've directly seen with Loïc, the problem is that if we drop it here, we will have to drop server-side merged commits too as it already implement this part of the protocol.

kerneis-anssi replied:
I'm fine with leaving things as-is.

nbouchinet-anssi commented on clients/go-mobile/create_archive.sh:
```diff
@@ -0,0 +1,26 @@
+
+if [ $# -eq 1 ]; then
+	echo "Copying gomobile.aac to $1"
+	cp gomobile.aar $1
```
shellcheck: Double quote to prevent globbing and word splitting.

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/database/DeviceRepository.kt:
```diff
@@ -27,4 +27,8 @@ class DeviceRepository(private val deviceDao: DeviceDao) {
+
+    fun setPCRs(device: Device, newPCRs: ByteArray) {
+        deviceDao.setPCRs(device.uid, newPCRs)
+    }
```
LGTM, can you add a "TODO" comment explaining the future uses of this function

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/DeviceListFragment.kt:
```diff
@@ -124,7 +123,7 @@ class DeviceListFragment : Fragment(R.layout.fragment_device_list), ItemClickLis
             is QRResult.QRSuccess -> {
                 if (isMACAddressValid(result.content.rawValue.trim())) {
-                    val device = Device(0, "", result.content.rawValue.trim(), byteArrayOf(), 0, byteArrayOf())
+                    val device = Device(0, "", result.content.rawValue.trim(), byteArrayOf(), 0, byteArrayOf(), byteArrayOf(), byteArrayOf())
```
Can't you overload the "Device" constructor to initialize an empty Device object instead of giving those byteArrayOf() args

nbouchinet-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt:
```diff
@@ -229,41 +236,56 @@ class ProtocolFragment : Fragment() {
+                        logger?.push(CLog("Attestation failure", false))
+                    }
+                    logger?.push(Log("Closing connection"))
+                    onDeviceDisconnectionCallback = {
```
Please comment this hack

nbouchinet-anssi commented on README.md:
````diff
@@ -12,16 +12,24 @@ The classical Ultrablue control flow consists of several steps:
+dracut --add "crypt ultrablue" /path/to/initrd --force
+```
+
+Note that those options are not persistant and **ultrablue** will be removed from your initramfs on its next generation. See the dracut.conf(5) man page for persistant configuration.
````
Typo persistant/persistent

Commit:
Rename ekenroll function to sendRegistrationData

Now, the client decides wether or not to send back a secret
to extend the 9th PCR. This implies that the server must
provide him a way to know if he must send one for later
attestations during enrollment. This is done through a
boolean value passed alongside the EK pub/cert in the first
enrollment message.
Thus, this message no longer contains only an EK, but also
other "registration data". To be consistent, this function
has been renamed.

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

kerneis-anssi commented on server/protocol.go:
```diff
@@ -201,13 +201,13 @@ func credentialActivation(ch chan []byte, tpm *attest.TPM) (*attest.AK, error) {
 func attestation(ch chan []byte, tpm *attest.TPM, ak *attest.AK) error {
 	logrus.Info("Getting anti replay nonce")
-	var nonce []byte
+	var nonce Bytestring
```
Ideally this should go in the previous commit. If you don't want to go through the hassle of splitting the commit (git makes it painful), please reword this commit message ("Remove usage indications at runtime") to add a line: "Also fix nonce encoding on reception"

buckwell-anssi replied:
I'll just edit the commit message because of the time constraint.

kerneis-anssi commented on clients/go-mobile/create_archive.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Can we please have a clients/go-mobile/README.md (and a reference to it in clients/android/README.md to explain how to build)?

buckwell-anssi replied:
Ok

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt:
```diff
@@ -233,37 +233,37 @@ class ProtocolFragment : Fragment() {
-                        CLog(
-                            "$tag doesn't fit in one packet: message size = ${prepended.size}",
-                            false
+            val btAdapter = getBluetoothAdapter()
```
Did you reindent on purpose, or is this a trick of your editor? If you're happy with the previous indentation, let's keep it to avoid adding whitespace that makes that diff less readable.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -48,15 +72,58 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
                 device.ekn = key.N
                 device.eke = key.E.toInt()
                 device.ekcert = key.Cert
+                logger?.push(Log("Registering device"))
```
This log message doesn't make a lot of sense: I would expect it to appear earlier, when the communication with the laptop starts (eg. before readMsg in REGISTRATION_READ). Having two log messages surrounding what is merely a UI update strikes me as overkill, unless I'm missing some underlying work performed by viewModel.insert.

buckwell-anssi replied:
viewModel.insert actually saves the device in the app database, and isn't UI related. This log informs the user that the new enrolled device is actually stored.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -48,15 +72,58 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+                encodedAttestationKey = message
+                logger?.push(Log("Generating credential challenge"))
+                val credentialBlob = Gomobile.makeCredential(device.ekn, device.eke.toLong(), encodedAttestationKey)
+                credentialActivationSecret = credentialBlob.secret
```
// Store the secret to validate it in the next stages

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -48,15 +72,58 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+                logger?.push(Log("Comparing received credential"))
+                if (decryptedCredential.Bytes.contentEquals(credentialActivationSecret)) {
+                    logger?.push(CLog("Credential matches the generated one", true))
+                    state += 1
```
Why do you need to explicitly increment state and resume here, but not in the other cases? If I understand correctly, it's because this case doesn't end in a read or write, because you have split two steps to improve readability regardless of the message exchange. The split is a good idea, but you should comment to explain the surprise:
// This step does not end in a message read or write, move explicitly to the next one

buckwell-anssi replied:
You understood correctly.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -48,15 +72,58 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+            }
+            ATTESTATION_SEND_NONCE -> {
+                rand.nextBytes(attestationNonce)
+                logger?.push(CLog("Generated anit replay nonce", true))
```
anit -> anti

kerneis-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,72 @@
+	return key
+}
+
+func MakeCredential(ekn []byte, eke int, encodedAP []byte) *CredentialBlob {
```
Document:  https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

kerneis-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/google/go-attestation/attest"
+)
+
+type CredentialBlob struct {
```
Document:  https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

kerneis-anssi commented on clients/go-mobile/create_archive.sh:
```diff
@@ -0,0 +1,26 @@
+
+if [ ! -v ANDROID_HOME ]; then
+	echo "Setting ANDROID_HOME to ~/Android/Sdk"
+	export ANDROID_HOME="/home/lfalkau/Android/Sdk"
```
/home/lfalkau -> ${HOME}

buckwell-anssi replied:
Oops

kerneis-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: 2022 ANSSI
```
I can't find where we check that the AK is a genuine TPM key rather than a fake one from a virtual TPM.

kerneis-anssi commented on clients/go-mobile/client.go:
```diff
@@ -0,0 +1,72 @@
+	return key
+}
+
+func MakeCredential(ekn []byte, eke int, encodedAP []byte) *CredentialBlob {
```
I'm confused why the AK (attestation KEY), which is called as such in the Android Code, is called AP (attestation parameters) here. Even more confusing is the `AK: ap` line below. I'm not saying you're doing anything wrong, it seems to match the naming convention of go-tpm, but it's confusing: can you please add to the function comment a note explaining what AP means and why AK==AP (or, if there's a subtle difference, what the difference actually is)?

buckwell-anssi replied:
There is effectively a subtle difference:
- Ak is an actual attestation key
- Activation parameters contains several information, such as the public key itself, but also what we call it's name, which (if I remember well) is a hash of it's public part and it's template.

I'll indeed add a comment or change variable names to make that less confusing

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -48,15 +72,58 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+            CREDENTIAL_ACTIVATION -> {
+                encodedAttestationKey = message
+                logger?.push(Log("Generating credential challenge"))
+                val credentialBlob = Gomobile.makeCredential(device.ekn, device.eke.toLong(), encodedAttestationKey)
```
makeCredential returns an invalid blob in case of error. We should check it here.

Maybe better, raise an exception instead and catch it?

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/database/AppDatabase.kt:
```diff
@@ -8,7 +8,7 @@ import androidx.room.RoomDatabase
 /*
```
Unrelated to this file, but you forgot to --signoff some of your commits, please rebase to make sure it's present.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt:
```diff
@@ -44,7 +39,7 @@ val ultrablueChrUUID: UUID = UUID.fromString("ebee1790-50b3-4943-8396-16c0b7231c
     The protocol itself is defined in the UltrablueProtocol.kt file.
  */
 class ProtocolFragment : Fragment() {
-    private var state = State.ENROLLMENT
```
I won't fight for it, but in general it's a good idea to use enumerations with meaningful names instead of booleans. Not sure why you changed it here.

https://abseil.io/tips/94 (written for C++ but applies equally well to Java)

buckwell-anssi replied:
Good advice!

buckwell-anssi replied:
In our case, I don't know it it is relevant, as this boolean is used inside the class function, and not passed to any exposed or external function.
I wrote it as an enum first because I thought the whole protocol state machine woud be written here, but I later decided to move it to the UltrablueProtocol class

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/Logger.kt:
```diff
@@ -75,17 +72,20 @@ class Logger(private var activity: MainActivity?, private var textView: TextView
 
-    fun last(): Log {
-        return logs.last()
+    private fun updateUI() {
```
Add a comment to explain what this does? (I have no clue)

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -109,24 +123,78 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+            }
+            PCRS_HANDLE -> {
+                try {
+                    logger?.push(Log("Getting PCRs"))
```
Getting PCRs, what does that mean? Where do we get them from? It sounds like we are reading them from the TPM, but they have already been sent at this point.

 Extracting PCRs from quote?

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -109,24 +123,78 @@ class UltrablueProtocol(private val activity: MainActivity, private var device:
+                        logger?.push(CLog("PCRs entries match the stored ones", true))
+                        AttestationResult(false, "")
+                    } else {
+                        // TODO: Need deeper investigation to determine which PCR changed and why.
```
A common pratice (at Google anyway) is to create a github issue when you have long-term TODOs to implement, and to reference it in the comment so that you can store more details in the issue and update it later:
// TODO(#42): ...

kerneis-anssi commented on server/protocol.go:
```diff
@@ -99,13 +99,16 @@ func parseAttestEK(ek *attest.EK) (EK, error) {
 		return EK{}, errors.New("Invalid key type:" + reflect.TypeOf(ek.Public).String())
 	}
-	var c = ek.Certificate.Raw
+	var c []byte = make([]byte, 1)
```
Add a comment to explain why we need this trick.

buckwell-anssi replied:
Idk why I created a byte slice of length 1, just checked and an empty byte slice works perfectly.
Fixed

kerneis-anssi reviewed (commented):
I'm not 100% done with the review yet (didn't have time for the last 3 comments) but here is a batch before going to lunch.

Commit:
Rename some state machine steps name to be consistent

As for server-side, now that the first message contains
additional data in addition to the EK pub/cert, we
renamed the steps with consistent names.

EK_READ => REGISTRATION_READ
EK_READ_HANDLE => REGISTRATION_READ_HANDLE

Also renamed the 'ek' variable to 'registrationData'

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
End the protocol on Android side (without sending back any secret)

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

kerneis-anssi commented on server/protocol.go:
```diff
@@ -216,16 +209,28 @@ func attestation(ch chan []byte, tpm *attest.TPM, ak *attest.AK) error {
 		return err
 	}
-	if atterr != nil {
+	if regerr.Err == true {
```
`== true` is redundant, `if regerr.Err {` is enough.

kerneis-anssi commented on server/protocol.go:
```diff
@@ -216,16 +209,28 @@ func attestation(ch chan []byte, tpm *attest.TPM, ak *attest.AK) error {
+	return nil
+}
+
+func response(ch chan []byte) error {
```
This is the most undescriptive name ever ;-) Can we please have a comment here?

kerneis-anssi commented on README.md:
```diff
@@ -12,16 +12,24 @@ The classical Ultrablue control flow consists of several steps:
 
 ### 1. Enrollment
 To enroll a phone as a verifier, start the server in enroll mode. This will display a QR code on the terminal. From the phone, run the client app, and tap the **+** icon on the top right corner to show a QR code scanner. On scan, an encrypted Bluetooth Low Energy channel will be established, and the enrollment will run automatically. Upon success, a device card will appear on the home page of the ...
+**Ultrablue** can extend your **TPM2 PCR9** using a randomly generated value at enroll time, this is usefull if you want to bind your disk encryption to **TPM2 sealing**, if configured as so, **ultrablue** will extend back the **PCR9** during boot-time only if the attestation is successfull and trusted. you'll then have to use the `--pcr-extend` option during enrollment.
```
Split the sentence with a dot after "TPM2 sealing".

kerneis-anssi commented on README.md:
```diff
@@ -12,16 +12,24 @@ The classical Ultrablue control flow consists of several steps:
 
 ### 1. Enrollment
 To enroll a phone as a verifier, start the server in enroll mode. This will display a QR code on the terminal. From the phone, run the client app, and tap the **+** icon on the top right corner to show a QR code scanner. On scan, an encrypted Bluetooth Low Energy channel will be established, and the enrollment will run automatically. Upon success, a device card will appear on the home page of the ...
+**Ultrablue** can extend your **TPM2 PCR9** using a randomly generated value at enroll time, this is usefull if you want to bind your disk encryption to **TPM2 sealing**, if configured as so, **ultrablue** will extend back the **PCR9** during boot-time only if the attestation is successfull and trusted. you'll then have to use the `--pcr-extend` option during enrollment.
```
"you'll then have to use the `--pcr-extend` option during enrollment."
This sentence makes no sense to me. Do you mean something like the following?

To make use of PCR extension, you need to add the `--pcr-extend` option, both during enrollment and attestation.

(Or do you? If you don't need it during attestation, then you should explicitly call it out:
... during enrollment, but *not* for attestation runs because the protocol will take care of signalling the requirement for extending the PCR.)

nbouchinet-anssi replied:
You need to add the `--pcr-extend` option only during enrollment, it asks the client to generate a random secret which will  then be extended into PCR9 for the first time. The user will then be able to seal a secret to this measurement. The secret is also stored client side.
On attestation success, the client will send back the secret to the server which will extend PCR9 again.
If the `--pcr-extend` is not asked during enrollment, no secret will be stored client side and no PCR9 extension will happen during attestation.

This pattern is usefull as the ultrablue systemd unit cannot be altered according to user configuration when installed in the initramfs for now.

nbouchinet-anssi replied:
I don't think this is necessary to specify that the `--pcr-extend` flag is not needed for attestation runs as those line are located in the enrollment  section of the documentation.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -149,6 +151,9 @@ class UltrablueProtocol(private val activity: MainActivity, private var enroll:
                 } catch (e: Exception) {
                     logger?.push(CLog("${e.message}", false))
+                    attestationResponse = AttestationResponse(true, byteArrayOf())
+                    state = RESPONSE
```
Ha, setting the state explicitly is more readable than += 1. Maybe you could do the same in other cases? Not a hard requirement though.

buckwell-anssi replied:
Done

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt:
```diff
@@ -160,36 +165,45 @@ class UltrablueProtocol(private val activity: MainActivity, private var enroll:
+        } else {
+            byteArrayOf()
+        }
+        rand.nextBytes(secret)
```
It works but I find confusing the fact that you essentially "push zero random bytes" in the non-extend case. I'd find the following clearer:
val secret = byteArrayOf()
if (extend) { secret = ByteArray(16). rand.nextBytes(secret) }

kerneis-anssi commented on server/protocol.go:
```diff
@@ -82,30 +84,34 @@ func recvMsg[T any](obj *T, ch chan []byte) error {
+	Cert      []byte // x509 key certificate (one byte set to 0 if none)
+	N         []byte // Raw public key bytes
+	E         int    // Public key exponent
+	PCRExtend bool   // Wether or not PCR_EXTENSION_INDEX must be extended on attestation success
```
typo: Whether

kerneis-anssi commented on server/testbed/README.md:
````diff
@@ -24,4 +25,17 @@ following dracut command:
 dracut --add ultrablue /path/to/initrd --force
 ```
 
+### Remote attested based disk decryption
````
Remote attested -> Remote-attestation

kerneis-anssi commented on server/utils.go:
```diff
@@ -28,6 +29,20 @@ func getTPMRandom(size uint16) ([]byte, error) {
+	defer rwc.Close()
+
+	err = tpm2.PCREvent(rwc, tpmutil.Handle(index), secret)
+	if err != nil {
```
nit: you could simplify the end of this function into just:
`return tpm2.PCREvent(rwc, ...)`

buckwell-anssi replied:
I don't know, I think it makes it clear it can fail.

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt:
```diff
@@ -61,6 +61,11 @@ class ProtocolFragment : Fragment() {
     private var onDeviceConnectionCallback: ((BluetoothGatt) -> Unit)? = null
     private var onMTUChangedCallback: (() -> Unit)? = null
     private var onServicesFoundCallback: (() -> Unit)? = null
+    private var gatt: BluetoothGatt? = null
```
Consistency: move this above with the other non-callback variables. Maybe remove blank line too (not sure, since with have a function body for this one).

kerneis-anssi commented on clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt:
```diff
@@ -249,25 +254,38 @@ class ProtocolFragment : Fragment() {
+                        logger?.push(CLog("Device disconnected", true))
+                    }
+                    gatt.disconnect()
+                    if (success && (enroll || device.secret.isNotEmpty())) {
```
Please comment to explain this condition (in particular device.secret.isNotEmpty is unclear to me).

Commit:
Add a README to explain how to use gomobile

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

kerneis-anssi commented:
I want the client to check that it speaks to a hardware TPM, ie. that the AK is signed with the public key of a known TPM manufacturer.

Les données à caractère personnel recueillies et traitées dans le cadre de cet échange, le sont à seule fin d’exécution d’une relation professionnelle et s’opèrent dans cette seule finalité et pour la durée nécessaire à cette relation. Si vous souhaitez faire usage de vos droits de consultation, de rectification et de suppression de vos données, veuillez contacter ***@***.*** Si vous avez reçu ce message par erreur, nous vous remercions d’en informer l’expéditeur et de détruire le message. The personal data collected and processed during this exchange aims solely at completing a business relationship and is limited to the necessary duration of that relationship. If you wish to use your rights of consultation, rectification and deletion of your data, please contact: ***@***.*** If you have received this message in error, we thank you for informing the sender and destroying the message.

kerneis-anssi commented:
Maybe it's the EK that is signed, not the EK.

Les données à caractère personnel recueillies et traitées dans le cadre de cet échange, le sont à seule fin d’exécution d’une relation professionnelle et s’opèrent dans cette seule finalité et pour la durée nécessaire à cette relation. Si vous souhaitez faire usage de vos droits de consultation, de rectification et de suppression de vos données, veuillez contacter ***@***.*** Si vous avez reçu ce message par erreur, nous vous remercions d’en informer l’expéditeur et de détruire le message. The personal data collected and processed during this exchange aims solely at completing a business relationship and is limited to the necessary duration of that relationship. If you wish to use your rights of consultation, rectification and deletion of your data, please contact: ***@***.*** If you have received this message in error, we thank you for informing the sender and destroying the message.

Commit:
Add some comments

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Remove redundant '== true'

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>

Commit:
Fix typo

Signed-off-by: Loic Buckwell <loic.buckwell@stagiaires.ssi.gouv.fr>
